### PR TITLE
drivers: display: ssd16xx: Add support for the ssd1677 EPD driver chip

### PR DESCRIPTION
**Pull request by nisargjhaveri**
Add support for the SSD1677 EPD driver chip with support for up to 960x680 pixel displays.

Tested with the Waveshare 4.26" 800x480 display with XIAO BLE board. I believe it is the same as Good Display GDEQ0426T82. Tested with the samples/drivers/display sample.

The SSD1677 requires x address to be full address, instead of the byte address used by SSD16XX. Added a new quirk to handle this.

The display requires a different GDO control flag as the panel layout might be different, add an option to set this.

The display also requires the scan direction for y axis to be reversed, add an option to set this as well.

Example devicetree overlay for the display:
```
&spi2 {
    cs-gpios = <&gpio0 3 GPIO_ACTIVE_LOW>;
};

/ {
    chosen {
        zephyr,display = &ssd1677;
    };

    mipi_dbi {
        compatible = "zephyr,mipi-dbi-spi";
        spi-dev = <&spi2>;
        dc-gpios = <&gpio0 29 GPIO_ACTIVE_HIGH>;
        reset-gpios = <&gpio0 2 GPIO_ACTIVE_LOW>;
        #address-cells = <1>;
        #size-cells = <0>;

        ssd1677: ssd1677@0 {
            compatible = "solomon,ssd1677";
            mipi-max-frequency = <4000000>;
            reg = <0>;
            busy-gpios = <&gpio0 5 GPIO_ACTIVE_HIGH>;

            width = <480>;
            height = <800>;

            tssv = <0x80>;
            softstart = [AE C7 C3 C0 80];
            gdo-flags = <0x02>;
            scan-y-reverse;

            full {
                border-waveform = <0x01>;
            };

            partial {
                border-waveform = <0x80>;
            };
        };
    };
};
```

### CODE REVIEW

File: drivers/display/ssd16xx.c [hunk 1]
```diff
@@ -623,8 +681,8 @@ static int ssd16xx_clear_cntlr_mem(const struct device *dev, uint8_t ram_cmd)
 	 * Clear unusable memory area when the resolution of the panel is not
 	 * multiple of an octet.
 	 */
-	if (config->height % EPD_PANEL_NUMOF_ROWS_PER_PAGE) {
-		panel_h += 1;
+	if (config->quirks->x_addr_in_bytes) { 
+		panel_h = (config->height / EPD_PANEL_NUMOF_ROWS_PER_PAGE) + (config->height % EPD_PANEL_NUMOF_ROWS_PER_PAGE ? 1 : 0);
```
Comment (danieldegrasse) [inline drivers/display/ssd16xx.c:685]: Why make a ternary statement here? We can just round the division up, right?

`panel_h = (config->height + (EPD_PANEL_NUMOF_ROWS_PER_PAGE - 1)) / EPD_PANEL_NUMOF_ROWS_PER_PAGE;`

Comment (nisargjhaveri) [reply drivers/display/ssd16xx.c:685]: Yes, we should be able to do that. I just kept it similar to what we were doing before. Will change.

File: drivers/display/ssd16xx.c [hunk 2]
```diff
@@ -105,6 +112,8 @@ struct ssd16xx_config {
 	uint16_t height;
 	uint16_t width;
 	uint8_t tssv;
+	uint8_t gdo_flags;
+	bool scan_y_reverse;
```
Comment (Snevzor) [inline drivers/display/ssd16xx.c:116]: Why is this property needed? Is there a specific use-case that requires it?

Comment (nisargjhaveri) [reply drivers/display/ssd16xx.c:116]: Yes, as in the example, the Waveshare 4.26" 800x480 display needs. As with the default scan modes, the displayed image is flipped. We need to use the data entry modes that has the Y direction reversed for the display to be correct.

File: drivers/display/ssd16xx.c [hunk 3]
```diff
@@ -1063,7 +1138,9 @@ static struct ssd16xx_quirks quirks_solomon_ssd1681 = {
 		.height = DT_PROP(n, height),				\
 		.width = DT_PROP(n, width),				\
 		.rotation = DT_PROP(n, rotation),			\
+		.scan_y_reverse = DT_PROP(n, scan_y_reverse),	\
 		.tssv = DT_PROP_OR(n, tssv, 0),				\
+		.gdo_flags = DT_PROP_OR(n, gdo_flags, 0),			\
```
Comment (danieldegrasse) [inline drivers/display/ssd16xx.c:1143]: Please keep the `\` characters aligned here across all lines

**Review by danieldegrasse — COMMENTED** [review]
Change itself looks ok to me- one formatting nit, and please address the errors highlighted by the compliance script. You should be able to see the specific lines in the Github UI when looking at the PR diff

File: drivers/display/ssd16xx.c (hunk 1 above)
Comment (nisargjhaveri) [reply drivers/display/ssd16xx.c]: Updated

File: drivers/display/ssd16xx.c (hunk 3 above)
Comment (nisargjhaveri) [reply drivers/display/ssd16xx.c]: Updated.

**Review by danieldegrasse — COMMENTED** [review]
Please do not add additional "fixup" commits to address reviewer requests- squash the commits making changes requested by reviewers back into your original commit, and force push over the PR branch. If you'd like to run clang format on the PR to silence those CI warnings that is ok, but please do so in a separate commit from where you have functional changes (it seems you already did this, just want to make sure it is clear this change *should not* be squashed into the commit adding functionality)

**Comment by nisargjhaveri** [comment]
@danieldegrasse Thanks for the suggestions. I've rebased and force pushed the commits.

**Review by danieldegrasse — COMMENTED** [review]
You'll need a commit body in both commits, that is going to fail the CI checks currently

**Comment by nisargjhaveri** [comment]
@danieldegrasse Updated again. Sorry for the troubles. Hope it makes sense this time? :)

**Review by Snevzor — COMMENTED** [review]
Dropping my review as I don't think I can add any value here. My initial comment was answered and all is clear to me.

**Comment by nisargjhaveri** [comment]
@danieldegrasse Could you have another look? I had to push more minor changes to fix the CI checks.

**Comment by danieldegrasse** [comment]
> @danieldegrasse Could you have another look? I had to push more minor changes to fix the CI checks.

Looks like there is still an issue. You can run this check locally (on a Linux system) with `./scripts/ci/check_compliance.py`, provided you install the python dependencies in `scripts/requirements-compliance.txt`

**Comment by nisargjhaveri** [comment]
Thanks for that! I pushed again, and it passes the failing check (gitlint) locally.

**Comment by nisargjhaveri** [comment]
Hi, @danieldegrasse. Is there anything else pending from my end? What needs to be done to merge this PR if it looks good?

**Review by danieldegrasse — DISMISSED** [review]
Thanks for the ping, LGTM

**Comment by nisargjhaveri** [comment]
@galak, @jfischer-no, could you please have a look as well? Let me know if this needs any more changes?

**Comment by andysan** [comment]
> @andysan Could you please take a look?

I will, but it'll be a couple of days. I can't promise I'll be able to find time before the weekend.

**Review by kartben — DISMISSED** [review]
LGTM

**Comment by danieldegrasse** [comment]
@jfischer-no could you take a look? If you are good with these changes I think we should go ahead and move this PR forwards

**Review by jfischer-no — CHANGES_REQUESTED** [review]
Do not "Run `clang format` on parts of `ssd16xx.c` to format some recent changes correctly." Keep you changes consistent with the lines around.

**Comment by nisargjhaveri** [comment]
@jfischer-no, I've update the PR with formatting the while file with clang-format. Please have a look if it makes sense?

**Comment by andysan** [comment]
Sorry for not writing up a review sooner. It has been exceptionally hard to find time the last few week.

Overall, I think this looks good and should be committed. I have put some comments  below, but they are minor nits that don't need to be addressed here.

Reviewing this, I realised just how confusing the current implementation is. If memory serves me right, the original driver always swapped rows and columns (some data sheets are less explicit about rows and columns and just refer to them as gates and sources which adds to the confusion) with respect to the data sheet since the most displays are used in landscape mode. 

When reviewing the code,  I started to think about ways of cleaning up the handling of rotation and inverted axes. The current code has an unfortunate amount of code duplication and feels slightly hard to follow. However, I haven't been able to come up a neater way of doing this. Annoyingly, with the exception of SSD1677, most other chips let you invert at least one of the axes in hardware.

As a general note, this PR contains a few slightly unrelated changes that could've been split into different commits. I would personally have split it into the following commits:


1. Clang format (first commit)
2. X address in bytes
3. Inverted y axis
4. SSD1677 support

That said,  I don't think it is forth the hassle of fixing up the commit history now. Again, sorry for taking such a long time.

**Comment by jfischer-no** [comment]
> @jfischer-no, I've update the PR with formatting the while file with clang-format. Please have a look if it makes sense?

No, it does not.

**Comment by jfischer-no** [comment]
> > > @jfischer-no, I've update the PR with formatting the while file with clang-format. Please have a look if it makes sense?
> > 
> > 
> > No, it does not.
> 
> This does not help much. Can you please suggest how should I approach this? :)
> 
> > Do not "Run clang format on parts of ssd16xx.c to format some recent changes correctly." Keep you changes consistent with the lines around.

Exactly as it is meant here. https://github.com/zephyrproject-rtos/zephyr/pull/75403#pullrequestreview-2461894121

> 
> If you meant, I should not run clang format at all but just be consistent, that does not seem to work as CI checks fails with the existing formatting in this file.

No, it should not be necessary to pass CI.

**Comment by nisargjhaveri** [comment]
Alright, removed the clang format commit. @jfischer-no

File: drivers/display/ssd16xx.c [hunk 4]
```diff
@@ -60,6 +60,14 @@ struct ssd16xx_quirks {
 	 * SSD16XX_CMD_UPDATE_CTRL2 for a partial refresh.
 	 */
 	uint8_t ctrl2_partial;
+
+	/*
+	 * Device specific flag deciding whether to pass in the byte address or the bit address
+	 * when setting the RAM x position window or counter in SSD16XX_CMD_RAM_XPOS_CTRL or
+	 * SSD16XX_CMD_RAM_XPOS_CNTR.
+	 * Expected to be `true` for the devices that expects the byte address, `false` otherwise.
+	 */
```
Comment (jfischer-no) [inline drivers/display/ssd16xx.c:69]: Please follow the style in `“nearby” code`.
```
        /*
         * Device specific flag deciding whether to pass in the byte address or
         * the bit address when setting the RAM x position window or counter in
         * SSD16XX_CMD_RAM_XPOS_CTRL or SSD16XX_CMD_RAM_XPOS_CNTR. Expected to
         * be `true` for the devices that expects the byte address, `false`
         * otherwise.
         */
```

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/display/ssd16xx.c [hunk 5]
```diff
@@ -588,17 +634,30 @@ static int ssd16xx_set_pixel_format(const struct device *dev,
...
+	} else {
+		if (orientation == DISPLAY_ORIENTATION_NORMAL) {
+			data->scan_mode = SSD16XX_DATA_ENTRY_XDYIY;
+		} else if (orientation == DISPLAY_ORIENTATION_ROTATED_90) {
+			data->scan_mode = SSD16XX_DATA_ENTRY_XDYDX;
+		} else if (orientation == DISPLAY_ORIENTATION_ROTATED_180) {
+			data->scan_mode = SSD16XX_DATA_ENTRY_XIYDY;
+		} else if (orientation == DISPLAY_ORIENTATION_ROTATED_270) {
+			data->scan_mode = SSD16XX_DATA_ENTRY_XIYIX;
+		}
```
Comment (IAmAnkur) [inline drivers/display/ssd16xx.c:660]: More of a suggestion:
```suggestion
		if (orientation == DISPLAY_ORIENTATION_NORMAL) {
			data->scan_mode = config->scan_y_reverse ? SSD16XX_DATA_ENTRY_XDYDY : SSD16XX_DATA_ENTRY_XDYIY;
		} else if (orientation == DISPLAY_ORIENTATION_ROTATED_90) {
			data->scan_mode = config->scan_y_reverse ? SSD16XX_DATA_ENTRY_XIYDX : SSD16XX_DATA_ENTRY_XDYDX;
		} else if (orientation == DISPLAY_ORIENTATION_ROTATED_180) {
			data->scan_mode = config->scan_y_reverse ? SSD16XX_DATA_ENTRY_XIYIY : SSD16XX_DATA_ENTRY_XIYDY;
		} else if (orientation == DISPLAY_ORIENTATION_ROTATED_270) {
			data->scan_mode = config->scan_y_reverse ? SSD16XX_DATA_ENTRY_XDYIX : SSD16XX_DATA_ENTRY_XIYIX;
		}
```
